### ISSUE TITLE
Comms plan for merging dagster changes into `dev`

### ISSUE THREAD

**Issue by jdangerx** [catalyst-cooperative/pudl#2295]
For context, the `dagster-asset-etl` branch does not touch the `PudlTabl` class. The branch mostly wraps our ETL in dagster concepts and does some reorganization of code and modules.

Answer the following questions:
* [ ] how will development work change once dagster is merged in?
* [ ] how will this affect people who are using forks?
* [ ] how will this affect people who are installing pudl from dev?
* [ ] how will we support people through this big transition?

**Comment by bendnorman** [comment]
Merging the `datgster-asset-etl` branch into `dev` will cause issues for folks who are installing the package from the dev branch and running the full ETL or specific extract/transform steps. I'm assuming not that many folks are in this situation. It seems like most people depend on the `PudlTabl` class. What do you think @zaneselvans @cmgosnell?

**Comment by jdangerx** [comment]
Just spent a bit of time thinking about this - here are some notes

How will this affect development? 

* merging dev will be a disaster. I think we'll want to take notes on the merge and then that will be a really useful guide.

* you will want to use dagster to make any new ETLs... don't think that needs tons of comms, I think most "new ETL development" lives with the cats, but maybe I'm wrong!

* pudl_etl etc. will still work, but be faster - we should tell people that it is now using dagster under the hood.

Beyond that - are there any particularly involved community members we should run this by / help out? 
* Kurt, Greg, and Alex have gotten PRs merged recently. Anyone else who we need to give a heads up to?

People who are installing off of dev should probably be ok as long as we keep the dagster dependency in `setup.py` - they won't get affected until they re-install off of pip, and then pip will pick up the new deps.

Things we can consider doing:
* [ ] take notes on the dev merge & publish them somewhere (a gist? then we can link to it all over social?)
* [ ] short blurb about "here's what's changing, it will be fun, I promise!"
* [ ] office hours for merge help? I guess we already have office hours?

**Comment by zschira** [comment]
A couple thoughts I have:

1. I think as much as is possible, we should try to get open branches merged into dev before the merge to minimize the number of times we have to resolve the same sorts of conflicts
2. Once we start working on the merge we should try to not stop until everything is in dev. If things continue to diverge while we're working on resolving conflicts this could drag on.
3. Once dagster is in dev everyone should immediately merge dev into any open branches they're working on (hopefully notes will help)
4. We should do a demonstration/walkthrough at least for our team

**Comment by jdangerx** [comment]
As noted in slack, we can also start merging `dev` into `dagster-asset-etl` or rebasing `dagster-asset-etl` onto `dev` basically the moment we have tests that pass. Once that is done, any conflicts from new work in `dev` should be much smaller...

**Comment by cmgosnell** [comment]
> Merging the datgster-asset-etl branch into dev will cause issues for folks who are installing the package from the dev branch and running the full ETL or specific extract/transform steps. I'm assuming not that many folks are in this situation. It seems like most people depend on the PudlTabl class. What do you think @zaneselvans @cmgosnell?

@bendnorman I expect that almost all users of the software do depend mostly on `PudlTabl` instead of the interim extract or transform functions directly. We've certainly attempted to corral folks into using `PudlTabl`.. but I'm also sure there are some off-brand uses out there that we are not aware of. I know @arengel used the extract step to get at more EIA data, but I believe it was not in an ongoing way and I think we've now incorporated at least some of that into dev.

Regardless, should we make a release from current dev or even main with a bunch of depreciation warnings for the things we are removing/renaming and such? We've generally thought of the database structure and `PudlTabl` as the thing that we are attempting to keep as stable as possible, but should we provide some warning to these potential ET users with warnings? I like a lot of the other ideas y'all have mentioned as well!

**Comment by zaneselvans** [comment]
In addition to @grgmiller, @arengel and @knordback I tink we probably want to give @gschivley a heads up about these changes, but I think he's still using an older SQLite DB, and will be happy to be able to just use a new DB file without the need to install any extra software, once we've got the output / analysis tables integrated into the DB.

Part of the reason we're making this transition is that most potential users found the software setup / dependencies prohibitively complex, so I think there aren't even that many people using the `PudlTabl` output objects, let alone running the ETL from scratch and intervening in its guts. And anybody who is doing that is probably going to be relatively technical.

For both internal usage and the enthusiastic outside contributor, I think we'll need a pretty big addition / revision to the "development setup" documentation -- e.g. what's the process for developing new table transforms like the ones that @e-belfer is working on in #2319? I imagine we'll need to replace or update the EIA ETL debugging notebook with a Dagster-centered process for running individual steps of the process on a subset of the years or tables.  But we're probably going to have to flesh this development setup / guide out as we go and learn what works. But if @zschira and @bendnorman and @jdangerx who have actually been doing data processing in Dagster already have an outline of what it'll look like that we can start with that would be great.

I suspect that one of the main consequences of this change will be an even bigger partitioning between the work that happens on the inside of the ETL, which is mostly done by us, and the usage of the data by a wider population of people.  That doesn't sound great, except that this split is already pretty stark, and as a result of this refactor, we're going to move a bunch of data out into a much more accessible place (the database, rather than output / analysis functions) so from the outside I think mostly people will see "Wow, look at all these useful new tables!" with only a few people thinking "Wow, look at all these new abstractions I don't understand."

Once we've finished migrating, how hard / easy to we think it **should** be to add a new asset? If someone develops a great new analysis or dataset of general use, can we provide a template for all the things that need to be defined so it shows up in the data warehouse / nightly builds?

**Comment by arengel** [comment]
The vast majority of the data that we use from PUDL we access through `PudlTabl` backed by SQLite DBs we download from datasette. We have poked around at the ETL but haven't used it in our pipeline. The only thing that might count as using the ETL is pulling CEMS data but that might be a separate process and not relevant to this discussion.

Also maybe not relevant to this discussion but adding it here just in case, we do you use functions from `analysis` and `helpers` directly so if those functions move or their APIs change, a round of deprecation warnings or some kind of guide to the changes would be helpful.

So long as the `PudlTabl` API doesn't change, I don't think we need any special consideration. If / when you want to guide us away from `PudlTabl`, I think we would want a mapping between `PudlTabl` methods and the new names of the tables in a SQLite DB or elsewhere as well as notice that it was happening.

Down the road I could see us wanting to contribute code to add new data to the ETL process so documentation to support that in this new system would be great, but this isn't a near term need.

**Comment by jdangerx** [comment]
So it sounds like we have a plan! Which is:

* [ ] tag a release right before we merge `dagster` into `dev` and then `dev` into `main`
* [ ] update docs:
  * [ ] how do we get the environment set up so we can run ETLs locally?
  * [ ] how to make an extraction asset?
  * [ ] how to make a transform asset?
  * [ ] how to write outputs of the extraction or transform assets?
  * [ ] how to make an asset that depends on transformed data?
  * [ ] how to materialize just one asset?
* [ ] tell people that the docs have been updated, that we are hosting office hours after the dev merge, and that we have a pre-dagster tag if people need it.

We can re-scope #1872 to just be "minimum viable documentation" which covers the above questions. Does that sound good @bendnorman @zschira @cmgosnell @zaneselvans ?

**Comment by bendnorman** [comment]
That plan sounds good to me! 

I wasn't planning on adding docs on how to add output assets until we've settled on an output table design.

**Comment by zaneselvans** [comment]
I guess I'm not really sure what fraction of the things we're doing day to day the above list of updates will cover.

The thing I'd really like to have as an example of how to work with the new system is a notebook or several notebooks, that work for developing new assets, or modifying / debugging existing parts of the data pipeline. One I frequently use is `devtools/eia-etl-debug.ipynb` but there are others under `devtools` too. I'm not sure which of them are up to date and commonly used by others. For me at least I think that will be more concretely helpful than just written documentation (though of course we should do that too). So maybe we should find out which of those people use and update them to work with the new system, which will be useful in its own right, but also provide concrete examples as a kind of live documentation.

**Comment by jdangerx** [comment]
New plan:
* [ ] tag release as above
* [ ] notify people as above
* [ ] update docs:
  * [ ] update `devtools/*debug.ipynb` (at least `eia-etl` and `ferc1-etl`, which I know people use - do people use `harvesting`?) to work in dagster world
  * [ ] environment setup documentation updates

**Comment by cmgosnell** [comment]
on the harvesting `devtools` notebook... it is rarely used but its super helpful in part bc it sets up a baby ET for EIA which spits out the stage before the gazillion options for all of the values which we are normalizing for the entities. Idk what the plan is in terms of implementing the interim-stage caching but in theory the need for this notebook should be obviated.

**Comment by bendnorman** [comment]
Sounds good! I've started to convert the `devtools/` notebooks in #2356. We'll have notebooks that run the baby ET steps and let you inspect the outputs of each step. 

I've also created another notebook that allows you to inspect asset values from the most recent ETL runs. This notebook can be used in tandem with dagit to execute portions of the ETL in dagit and inspect them in the notebook for debugging.

**Comment by jdangerx** [comment]
Sounds like we have a plan! @bendnorman let me know if you want me to take one of the remaining notebooks in #2356. @zaneselvans pipe up if you feel this plan didn't take your feedback into account correctly!

**Comment by zaneselvans** [comment]
I think this is good enough. I'm sure it'll impact someone, but hopefully only relatively savvy folks. Mostly I think we need to make sure that we communicate to ourselves, and write up whatever we learn in our own process switch, which should provide a  richer, more informed set of documentation for others.
